File: Ywy/JAVA/src/level2/TargetNumber.java
package level2;

public class TargetNumber {
	private int target;
	private int range;
	private int[] numbers;
	public int solution(final int[] numbers, final int target) {
		this.target = target;
		this.range = numbers.length;
		this.numbers = numbers;
		return dfs(0, 0);
    }

    private int dfs(int depth, int sum) {
        if (depth == range) {
            if (sum == target) {
                return 1;
            }
            return 0;
        }
        int el = numbers[depth++];
        return dfs(depth, sum + el) + dfs(depth, sum - el);
    }
}

File: Ywy/JavaScript/level2/TargetNumber.js
export default class TargetNumber {
    solution = (numbers, target) => {
        const range = numbers.length;
        const dfs = (depth, sum) => {
            if (depth === range) {
                return sum === target;
            }
            let el = numbers[depth];
            return dfs(++depth, sum + el) + dfs(depth, sum - el);
        };
        return dfs(0, 0);
    };
}